javalib/src/main/scala/java/net/URI.scala

```diff
@@ -106,36 +106,45 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
       // Undefined components are considered less than defined components
       else x.fold(-1)(s1 => y.fold(1)(s2 => cmp(s1, s2)))
     }
+    @inline def cmpOptIgnoreCase(x: js.UndefOr[String], y: js.UndefOr[String]): Int = {
+      if (x == y) 0
+      // Undefined components are considered less than defined components
+      else x.fold(-1)(s1 => y.fold(1)(s1.compareToIgnoreCase))
+    }
 
-    if (this._scheme != that._scheme)
-      this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
-    else if (this._isOpaque != that._isOpaque)
-      // A hierarchical URI is less than an opaque URI
-      if (this._isOpaque) 1 else -1
-    else if (_isOpaque) {
+    @inline def cmpSsp(): Int = {
       val ssp = cmp(this._schemeSpecificPart, that._schemeSpecificPart)
       if (ssp != 0) ssp
       else cmpOpt(this._fragment, that._fragment)
-    } else if (this._authority != that._authority) {
+    }
+
+    @inline def cmpPathQueryFragment(): Int = {
+      if (this._path != that._path)
+        cmpOpt(this._path, that._path)
+      else if (this._query != that._query)
+        cmpOpt(this._query, that._query)
+      else
+        cmpOpt(this._fragment, that._fragment)
+    }
+
+    List(
+      cmpOptIgnoreCase(this._scheme, that._scheme),
+      this.isOpaque.compareTo(that.isOpaque), // A hierarchical URI is less than an opaque URI
+      if (this.isOpaque) cmpSsp() else 0,
+      if (this._authority == that._authority) cmpPathQueryFragment() else 0,
       if (this._host.isDefined && that._host.isDefined) {
         val ui = cmpOpt(this._userInfo, that._userInfo)
         if (ui != 0) ui
         else {
           val hst = this._host.get.compareToIgnoreCase(that._host.get)
           if (hst != 0) hst
-          else if (this._port == that._port) 0
+          else if (this._port == that._port) cmpPathQueryFragment()
           else if (this._port == -1) -1
           else if (that._port == -1)  1
           else this._port - that._port
         }
-      } else
-        cmpOpt(this._authority, that._authority)
-    } else if (this._path != that._path)
-      cmpOpt(this._path, that._path)
-    else if (this._query != that._query)
-      cmpOpt(this._query, that._query)
-    else
-      cmpOpt(this._fragment, that._fragment)
+      } else cmpOpt(this._authority, that._authority)
+    ).find(_ != 0).getOrElse(0)
```

Comment:
I think the current version is kind of hard to read.

How about something like this:

```scala
List(
    cmpOpt(this._scheme, that._scheme)(_.compareToIgnoreCase),
    this.isOpaque.compareTo(that.isOpaque), // A hierarchical URI is less than an opaque URI
) ++ (
    if (this.isOpaque) {
      List(
        cmp(this._schemeSpecificPart, that._schemeSpecificPart),
      )
    } else if (this._host.isDefined && that._host.isDefined) {
      List(
        cmpOpt(this._userInfo, that._userInfo)(cmp),
        cmpOpt(this._host, that._host)(_.compareToIgnoreCase),
        cmpOpt(this._port, that._port)(_ - _),  // change _port to js.UndefOr[Int] and fold into -1 only when getPort is called.
      )
    } else {
      List(
         cmpOpt(this._authority, that._authority)(cmp)
      )
    }
) ++ List(
    cmpOpt(this._path, that._path)(cmp),
    cmpOpt(this._query, that._query)(cmp),
    cmpOpt(this._fragment, that._fragment)(cmp),
).find(_ != 0).getOrElse(0)
```

It is unfortunately not as nice as I anticipated, mostly because of the if-branches. Not sure how to improve though. Maybe something with `Option#orElse` or helper functions would work better.

Reply:
Refactoring so and cleanup commits.